performance/multi-node/random_access.c

```diff
@@ -0,0 +1,1081 @@
+		ran = (ran << 1) ^ ((int64_t)ran < ZERO64B ? POLY : ZERO64B);
+		global_offset = ran & (table_size - 1);
+		if (global_offset < top) {
+			remote_pe = global_offset / (min_local_table_size + 1);
+			global_start_at_pe = (min_local_table_size + 1) * remote_pe;
+		} else {
+			remote_pe = (global_offset - remainder) / min_local_table_size;
+			global_start_at_pe = min_local_table_size * remote_pe + remainder;
+		}
+		index = global_offset - global_start_at_pe;
+
+		lf_atomic_xor((void *)&table[index], ran, remote_pe);
+	}
+
+	ctpm_Barrier();
+}
+
+static int random_access(void)
+{
+	int64_t i;
+	static int64_t num_errors, global_num_errors;
+	int remainder; /* Number of processors with (LocalTableSize + 1) entries */
+	uint64_t num_top_table_entries; /* Number of table entries in top of Table */
+	int64_t local_table_size; /* Local table width */
+	uint64_t min_local_table_size; /* Integer ratio TableSize/num_ranks */
+	uint64_t log_table_size;
+	uint64_t table_size;
+	double real_time; /* Real time to update table */
+	double total_mem;
+	static int send_abort, recv_abort;
+	uint64_t default_num_updates; /* Number of updates to table
+	 (suggested: 4x number of table entries) */
+	uint64_t actual_num_updates; /* actual number of updates to table - may be
+	 smaller than NumUpdates_Default due to
+	 execution time bounds */
+	int64_t proc_num_updates; /* number of updates per processor */
+	FILE *outFile = NULL;
+	double MyGUPS;
+	double *local_gups_score;
+	int rc;
+
+	local_gups_score = &MyGUPS;
+
+	if (HEAD_RANK) {
+		outFile = stdout;
+		setbuf(outFile, NULL);
+	}
+
+	/*
+	 * TODO: replace this
+	 */
+
+	total_mem = tunables.total_mem_opt ? tunables.total_mem_opt : 200000; /* max single node memory */
+	total_mem *= num_ranks; /* max memory in num_ranks nodes */
+
+	total_mem /= sizeof(uint64_t);
+
+	/* calculate TableSize --- the size of update array (must be a power of 2) */
+	for (total_mem *= 0.5, log_table_size = 0, table_size = 1; total_mem >= 1.0;
+			total_mem *= 0.5, log_table_size++, table_size <<= 1)
+		; /* EMPTY */
+
+	/*
+	 * Calculate local table size, etc.
+	 */
+	min_local_table_size = table_size / num_ranks;
+
+	/* Number of processors with (LocalTableSize + 1) entries */
+	remainder = table_size - (min_local_table_size * num_ranks);
+
+	/* Number of table entries in top of Table */
+	num_top_table_entries = (min_local_table_size + 1) * remainder;
+
+	/* Local table size */
+	if (rank < remainder) {
+		local_table_size = min_local_table_size + 1;
+		global_start_my_proc = ((min_local_table_size + 1) * rank);
+	} else {
+		local_table_size = min_local_table_size;
+		global_start_my_proc = ((min_local_table_size * rank) + remainder);
+	}
+
+	send_abort = 0;
+	hpcc_table = malloc(local_table_size * sizeof(uint64_t));
+	if (!hpcc_table)
+		send_abort = 1;
+
+	hpcc_lock = malloc(sizeof(uint64_t) * num_ranks);
+	if (!hpcc_lock)
+		send_abort = 1;
+
+	ctpm_Barrier();
+	ctpm_int_sum_all(&recv_abort, num_ranks);
```

Comment:
recv_abort is the destination address for aggregating all of the send_abort values.